### PR TITLE
smartcontract: redesign reservation as block with count and add reserved subscribe instructions

### PR DESCRIPTION
**Pull request by snormore**
## Summary of Changes
- Redesign reservation from a single-IP model to a block-based model with a `count` field, allowing one reservation to hold multiple seats. PDA is now keyed by `(device, owner)` instead of `(device, client_ip)`.
- Add `CreateReservedSubscribeUser` instruction that consumes a reservation seat to create a multicast subscriber without an access pass — authorized via reservation authority or foundation allowlist.
- Add `DeleteReservedSubscribeUser` instruction for tearing down reserved users (unsubscribe, deallocate resources, close account).
- Add multicast capacity checking (`max_multicast_users`) to the reserve instruction so seats can't be reserved beyond the per-type limit.
- Use checked arithmetic throughout reservation processors to prevent overflow/underflow.

## Diff Breakdown
| Category     | Files | Lines (+/-) | Net   |
|--------------|-------|-------------|-------|
| Core logic   |    10 |  +146 / -77 |  +69  |
| New instructions | 2 |  +517 / -0  | +517  |
| Tests        |     1 | +981 / -68  | +913  |
| Config/build |     1 |   +1 / -0   |   +1  |

~65% of this PR is test code; core logic adds two new instructions plus reservation redesign.

<details>
<summary>Key files (click to expand)</summary>

- [`create_reserved_subscribe.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-a7fe9be32f18a95760aebbee2af66be36dfb5585d43138a314b409129ded234b) — New `CreateReservedSubscribeUser` instruction: creates multicast user from reservation, subscribes to group, optional onchain resource allocation
- [`delete_reserved_subscribe.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-1061e9af954b694f71a8cf265ef2658cc0b3e0efb27aacbd30bb80fc3e684f10) — New `DeleteReservedSubscribeUser` instruction: unsubscribes, deallocates resources, closes user account
- [`reservation_test.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-edac94c6aee34fb47dfc15c6c455d20ec869756e8df2e180351aabcefa0df732) — Extensive test coverage for block reservations, create/delete reserved subscribe, capacity checks, edge cases
- [`reserve.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-4b886b3cec2101806a8de60e9ff4100edd4fa80c4badfcedbc777c7594efe57a) — Redesigned: `count` replaces `client_ip`, adds multicast capacity check, checked arithmetic
- [`resource_onchain_helpers.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-c29105e0399b62fe0f69ee2a5f5876e73b91e9e4366daf15bba27be17da21b90) — Extract `parse_resource_extension_accounts` helper shared by create instructions
- [`reservation.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-8d79c5ef9dbb6036b7a4b77be8160d0d718df788a31a3cefaade8f812a78b4f1) — Replace `client_ip: Ipv4Addr` with `reserved_count: u16`
- [`instructions.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-3aed842c99767fc5c576e142220dab61652c1872588c96ccbe1476ec39b8db3d) — Register variants 102/103 for new instructions
- [`close.rs`](https://github.com/malbeclabs/doublezero/pull/3224/files#diff-a41afb2ac00f5d20e0c7e87357a4818caa53bd8f05d8b67edc07ad769eaf879b) — Decrement `reserved_seats` by remaining `reserved_count` on close

</details>

## Testing Verification
- Reservation tests cover: block reservation with count, multicast capacity enforcement, close returning correct seat count, overflow/underflow protection
- CreateReservedSubscribeUser tests: happy path with onchain allocation, non-multicast rejection, reservation consumption and exhaustion, authorization checks
- DeleteReservedSubscribeUser tests: full teardown with resource deallocation, multicast group counter updates, device counter decrements

### CODE REVIEW

File: smartcontract/programs/doublezero-serviceability/src/processors/reservation/reserve.rs [hunk 1]
```diff
@@ -79,15 +83,19 @@ pub fn process_reserve_connection(
...
 
-    // Check device capacity: users_count + reserved_seats must be less than max_users.
+    // Check device capacity: users_count + reserved_seats + count must not exceed max_users.
     // A device with max_users == 0 is locked and cannot accept reservations.
-    if device.users_count + device.reserved_seats >= device.max_users {
+    let new_reserved = device
+        .reserved_seats
+        .checked_add(value.count)
+        .ok_or(DoubleZeroError::MaxUsersExceeded)?;
+    if device.users_count + new_reserved > device.max_users || device.max_users == 0 {
```
[the remark this answers is not in the document]
Comment (karl-dz) [reply smartcontract/programs/doublezero-serviceability/src/processors/reservation/reserve.rs]: Also is checking that the max == 0 necessary? I think the inequality would be violated anyway

File: smartcontract/programs/doublezero-serviceability/src/processors/reservation/reserve.rs [hunk 2]
```diff
@@ -79,15 +83,23 @@ pub fn process_reserve_connection(
...
-    if device.users_count + device.reserved_seats >= device.max_users {
+    let new_reserved = device
+        .reserved_seats
+        .checked_add(value.count)
+        .ok_or(DoubleZeroError::MaxUsersExceeded)?;
+    let total_occupied = device
+        .users_count
+        .checked_add(new_reserved)
+        .ok_or(DoubleZeroError::MaxUsersExceeded)?;
+    if total_occupied > device.max_users || device.max_users == 0 {
```
[the remark this answers is not in the document]
Comment (karl-dz) [reply smartcontract/programs/doublezero-serviceability/src/processors/reservation/reserve.rs]: Also note: if the max multicast users is zero, then we should check max users (since max multicast users == 0 disables the multicast count tracking)
